video/out/vo_gpu_next.c

```diff
@@ -1515,84 +1521,216 @@ static void wait_events(struct vo *vo, int64_t until_time_ns)
     }
 }
 
-#if PL_API_VER < 342
-static inline void xor_hash(void *hash, pl_cache_obj obj)
+static char *cache_filepath(void *ta_ctx, char *dir, const char *prefix, uint64_t key)
 {
-    *((uint64_t *) hash) ^= obj.key;
+    bstr filename = {0};
+    bstr_xappend_asprintf(ta_ctx, &filename, "%s_%016" PRIx64, prefix, key);
+    if (!filename.len)
+        return NULL;
+    return mp_path_join_bstr(ta_ctx, bstr0(dir), filename);
 }
 
-static inline uint64_t pl_cache_signature(pl_cache cache)
+static void cache_save_file(void *ta_ctx, char *filepath, void *data, size_t size)
 {
-    uint64_t hash = 0;
-    pl_cache_iterate(cache, xor_hash, &hash);
-    return hash;
+    if (!data || !size)
+        return;
+
+    char *tmp = talloc_asprintf(ta_ctx, "%sXXXXXX", filepath);
+    int fd = mkstemp(tmp);
+    if (fd < 0)
+        return;
+    FILE *cache = fdopen(fd, "wb");
+    if (!cache) {
+        close(fd);
+        unlink(tmp);
+        return;
+    }
+    size_t written = fwrite(data, size, 1, cache);
+    int ret = fclose(cache);
+    if (written > 0 && !ret) {
+        ret = rename(tmp, filepath);
+    } else {
+        unlink(tmp);
+    }
+}
+
+static pl_cache_obj cache_load_obj(void *p, uint64_t key)
+{
+    struct cache *c = p;
+    void *ta_ctx = talloc_new(NULL);
+    pl_cache_obj obj = {0};
+
+    if (!c->dir)
+        goto done;
+
+    char *filepath = cache_filepath(ta_ctx, c->dir, c->name, key);
+    if (!filepath)
+        goto done;
+
+    if (stat(filepath, &(struct stat){0}))
+        goto done;
+
+    int64_t load_start = mp_time_ns();
+    struct bstr data = stream_read_file(filepath, ta_ctx, c->global, 1000000000);
```

Comment:
surely it doesn't really need to be 1GB?

Reply:
Good catch, indeed we need 1GiB per object to support those 512x512x512 3dluts, I removed this arbitrary limit and changed to INT_MAX.

I copied this line form vo_gpu in fact which also uses 1GiB, which is in fact no max size limit.

Reply:
so how big is a 512³ LUT exactly?

having a single object bigger than the hardcoded cache limit doesn't even make sense. it will get deleted every 24h.

Reply:
> so how big is a 512³ LUT exactly?

1GiB

> having a single object bigger than the hardcoded cache limit doesn't even make sense. it will get deleted every 24h.

That's why limit is 1.5GiB

Reply:
Then I suggest using `cache->size_limit ? MPMIN(2000000000, cache->size_limit) : 2000000000` as read limit.

And in `cache_save_obj` maybe even refuse to save files bigger than this.